### ISSUE TITLE
Eventlist: Removal of first/last classes leads to reduced functionality

### ISSUE THREAD

**Issue by bauer00** [contao/contao#6402]
### Affected version(s)

5.2.2

### Description

With Contao 5 you removed several CSS classes: https://github.com/contao/contao/blob/5.x/UPGRADE.md#css-classes-first-last-even-and-odd

In the case of the event list the suggested workaround using CSS selectors cannot provide the same functionality.

Steps to reproduce:
1. Select the module template "event_list" in the module type eventlist
2. Create 2 events on the same day
3. The rendered HTML looks like this: 
![grafik](https://github.com/contao/contao/assets/3669176/e8550d87-0c84-40d0-8859-575aa49db034)
4. The div selected in the screenshot (cursor) can't be selected with CSS selectors. In previous Contao versions this element would have a CSS class "last". (first and last classes are provided by day and not for the whole eventlist)

Is there a workaround for this, or is there a proposed replacement?

**Comment by ausi** [comment]
> Is there a workaround for this, or is there a proposed replacement?

`.first` is the same as `.header + .event`
`.last` is the same as `.event:has(+.header), .event:last-of-type`

But be aware that `:has(…)` is [not supported in all browsers](https://caniuse.com/css-has) yet.

**Comment by asaage** [comment]
I think the Template composition is just a bit unlucky here with unpredictable ```.header``` sneaking into ```.mod_eventlist```.

> `.first` is the same as `.header + .event`

will give you basically multiple "first events" depending on how many headers there are.
In some cases you might want that and in some not.
I dont' think you can target odd/even events if there is anything other than ```.event``` as child-elements of the list.

**Comment by ausi** [comment]
> will give you basically multiple "first events" depending on how many headers there are.

Same as it was in Contao 4 with the `.first` class, right?

> In some cases you might want that and in some not.

But in this ticket @bauer00 asked explicitly for this case (as it was in Contao 4).

> I dont' think you can target odd/even events if there is anything other than `.event` as child-elements of the list.

Why not? Should be no problem with `.event:nth-child(odd of .event)` and `.event:nth-child(even of .event)`

**Comment by asaage** [comment]
> Same as it was in Contao 4 with the `.first` class, right?

Honestly: I don't know - for me it looks weired.

> > I dont' think you can target odd/even events if there is anything other than `.event` as child-elements of the list.
> 
> Why not? Should be no problem with `.event:nth-child(odd of .event)` and `.event:nth-child(even of .event)`

That would treat the whole list (alternating odd/even events) and you can counterintuitively end up with first, even events.
I didn't run into issues with that yet - just saying it's a bit quirky.

**Comment by bauer00** [comment]
Thanks for your responses.

> `.first` is the same as `.header + .event` `.last` is the same as `.event:has(+.header), .event:last-of-type`

Yes, we also found this, but unfortunately it's not supported by Firefox at all, so it's not usable for us.

> But in this ticket @bauer00 asked explicitly for this case (as it was in Contao 4).

Yes, we try to group all events from a day together, which I think is not that unusual to do. If all events from a day would be nested into another element, using `:first-child` and `:last-child` would be possible again.

Our layout looks like this at the moment, but apart from a really bad workaround where we add additional div's in multiple templates, we weren't able to create the same layout with Contao 5.

![grafik](https://github.com/contao/contao/assets/3669176/16347c15-1d96-4b36-b0f4-7a8bbd887165)

**Comment by fritzmg** [comment]
I agree that this can be cumbersome. When dealing with layouts that show the "header" I usually adjust the event template like this:

```php
<?php

$eventClasses = explode(' ', $this->classList);
$isLastEvent = \in_array('last', $eventClasses);

?>
<?php if ($this->header): ?>
  <div class="wrapper">
    <div class="header"><?= $this->header ?></div>
    <div class="events">
<?php endif; ?>

<div class="event">
  <!-- … -->
</div>

<?php if ($isLastEvent): ?>
    </div>
  </div>
<?php endif; ?>
```

This wouldn't work anymore in Contao 5. I think https://github.com/contao/contao/pull/4178 was overzealous after all. For the event list these classes can still be useful (as shown by the other comments here) as first/last/even/odd has special meaning here.

**Comment by ausi** [comment]
As discussed in the call today, we want to set `$this->groupIndex` and `$this->groupCount` for the `event_*` templates to make it possible to handle all cases that where previously possible using the CSS classes.

`$this->groupIndex === 0` would be true for the first element
`$this->groupIndex === $this->groupCount - 1` would be true for the last element
